File: editor/en/[623]Add One Row to Tree.py
# leetcode submit region begin(Prohibit modification and deletion)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def addOneRow(self, root: Optional[TreeNode], val: int, depth: int) -> Optional[TreeNode]:
        if depth == 1:
            new_root = TreeNode(val)
            new_root.left = root
            return new_root

        self.insert_at_depth(root, val, depth)
        return root

    def insert_at_depth(self, node: Optional[TreeNode], val: int, depth: int) -> None:
        if node is None:
            return

        if depth == 2:
            temp_left = node.left
            temp_right = node.right
            node.left = TreeNode(val, temp_left, None)
            node.right = TreeNode(val, None, temp_right)
        else:
            self.insert_at_depth(node.left, val, depth - 1)
            self.insert_at_depth(node.right, val, depth - 1)
# leetcode submit region end(Prohibit modification and deletion)

File: editor/en/[623]Add One Row to Tree.java
//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    public TreeNode addOneRow(TreeNode root, int val, int depth) {
        if (depth == 1) {
            TreeNode newRoot = new TreeNode(val);
            newRoot.left = root;
            return newRoot;
        }

        insertAtDepth(root, val, depth);
        return root;
    }

    private void insertAtDepth(TreeNode node, int val, int depth) {
        if (node == null) {
            return;
        }

        if (depth == 2) {
            TreeNode tempLeft = node.left;
            TreeNode tempRight = node.right;
            node.left = new TreeNode(val, tempLeft, null);
            node.right = new TreeNode(val, null, tempRight);
        } else {
            insertAtDepth(node.left, val, depth - 1);
            insertAtDepth(node.right, val, depth - 1);
        }
    }
}
//leetcode submit region end(Prohibit modification and deletion)
